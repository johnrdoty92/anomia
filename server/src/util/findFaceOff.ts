import { Card } from "@prisma/client";

export function findFaceOff<T extends { card: Card | null }>(items: T[]): [T, T] | null {
  const hashmap = new Map<number, T>();
  for (const item of items) {
    if (!item.card) continue;
    const existingEntry = hashmap.get(item.card.shape1);
    if (existingEntry != null) return [existingEntry, item];
    // TODO: handle wild cards
    hashmap.set(item.card.shape1, item);
  }
  return null;
}
