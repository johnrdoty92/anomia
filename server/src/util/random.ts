const ALPHABET_RANGE = 26;
const ASCII_OFFSET = 65;

export const getGameId = () =>
  Array(6)
    .fill("")
    .map(() =>
      String.fromCharCode(
        Math.floor(Math.random() * ALPHABET_RANGE) + ASCII_OFFSET,
      ),
    )
    .join("");

export const getDeckSeed = () => Math.floor(Math.random() * 10000);