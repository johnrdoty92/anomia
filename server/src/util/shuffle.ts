// Taken from https://stackoverflow.com/a/47593316
function mulberry32(a: number) {
  return function () {
    let t = (a += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export function shuffle<T>({ items, seed }: { items: T[]; seed: number }) {
  const random = mulberry32(seed);
  const cardsCopy = structuredClone(items);
  for (let i = cardsCopy.length - 1; i > 0; i--) {
    const j = Math.floor(random() * i);
    const temp = cardsCopy[i];
    cardsCopy[i] = cardsCopy[j];
    cardsCopy[j] = temp;
  }
  return cardsCopy;
}
