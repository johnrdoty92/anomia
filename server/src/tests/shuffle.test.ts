import { describe, expect, it } from "vitest";
import { shuffle } from "../util/shuffle";

describe("shuffle", () => {
  it("should match the snapshot", () => {
    const length = 100;
    const shuffled = shuffle({
      items: Array(length)
        .fill(0)
        .map((_, i) => i),
      seed: 1,
    });
    expect(shuffled).toMatchSnapshot();
    expect(shuffled).toHaveLength(length);
    expect(new Set(shuffled).size).toBe(length);
  });
});
