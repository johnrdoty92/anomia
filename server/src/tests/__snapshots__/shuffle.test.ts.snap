// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shuffle > should match the snapshot 1`] = `
[
  82,
  17,
  20,
  7,
  98,
  86,
  53,
  97,
  64,
  79,
  42,
  45,
  39,
  27,
  49,
  60,
  65,
  41,
  80,
  84,
  54,
  23,
  69,
  57,
  96,
  74,
  77,
  25,
  99,
  37,
  73,
  67,
  16,
  75,
  10,
  30,
  1,
  4,
  35,
  76,
  36,
  28,
  70,
  93,
  92,
  68,
  18,
  33,
  47,
  11,
  88,
  55,
  2,
  58,
  6,
  81,
  83,
  8,
  72,
  24,
  19,
  9,
  71,
  15,
  63,
  50,
  29,
  48,
  13,
  52,
  95,
  46,
  78,
  85,
  59,
  44,
  32,
  3,
  14,
  22,
  61,
  31,
  5,
  90,
  87,
  21,
  34,
  12,
  43,
  40,
  89,
  38,
  66,
  56,
  26,
  91,
  94,
  51,
  0,
  62,
]
`;
